### ISSUE TITLE
Why is the BN layer not added before the softmax function?

### ISSUE THREAD

**Issue by daigang896** [voldemortX/pytorch-auto-drive#51]
Hello：
pytorch-auto-drive/torchvision_models/common_models.py         341 line: output = F.softmax(output, dim=1)。

Hello, why didn't you add the BN layer before the softmax function?  In addition, does this "class EDLaneExist(nn.Module) " have any optimization space when the maximum number of lanes exceeds 6?

**Comment by voldemortX** [comment]
@daigang896 Actually if you are considering optimization, the problem may be from the 5 set [here](https://github.com/voldemortX/pytorch-auto-drive/blob/b21cb07656a2bbd32f3cc4340422c9510badbc74/torchvision_models/common_models.py#L319). But the effects seem minor on current results (only some 6 lanes on TuSimple) so we kept it for BC.

What do you mean by BN after softmax, is it a common practice? The intuition here is to infer existence based on pixel-wise probability map.

**Comment by daigang896** [comment]
@voldemortX Hello, I'm just adding BN layer before this softmax, I'm not sure if it's useful.  Trying to see if it works. “pytorch-auto-drive/blob/master/torchvision_ models/common_ models. py:319: self.layers_ final. append(nn.Conv2d(32, 5, (1, 1), stride=1, padding=(0, 0), bias=True))". I set the number 5 a little bigger. I understand the intention here. I'm just thinking about how to improve the lane line accuracy a little more.

**Comment by voldemortX** [comment]
@daigang896 Sounds interesting, please inform me here if it works.

**Comment by daigang896** [comment]
@voldemortX OK.

**Comment by daigang896** [comment]
@voldemortX  It can not improve the accuracy of lane line detection. Maybe erfnet_SCNN or erfnet_RESA can only achieve a certain accuracy. The backbone network may need to be replaced.

**Comment by voldemortX** [comment]
@daigang896  Yes I agree that the bottleneck is the algorithms themselves. We are currently exploring alternative backbones such as RepVGG, MobileNet, and Transformers. Part of RepVGG might show up along with the upcoming refactoring (it's scheduled for 12.31 but we might not make it, just stay tuned).

**Comment by daigang896** [comment]
@voldemortX OK, look forward to new updates.

**Comment by voldemortX** [comment]
@daigang896 RepVGG, MobileNet, and Transformers are now supported (baselines and some RESA/SCNN). Furthermore, BézierLaneNet (CVPR 22) is integrated, which requires both keypoint and binary seg annotations (it should be easy to get them from keypoint annotations). However, we don't see a much higher performance on public datasets. You can try them out yourself now with the latest code.

I'll close this issue for now. Feel free to reopen or open a new one for further discussions.
